model/transformers/dffml_model_transformers/qa/qa_model.py

```diff
@@ -0,0 +1,1006 @@
+            is_training=False,
+            return_dataset="pt",
+        )
+
+        results = {}
+        if self.parent.config.local_rank in [-1, 0]:
+            logger.info(
+                "Loading checkpoints saved during training for evaluation"
+            )
+            self.model = AutoModelForQuestionAnswering.from_pretrained(
+                self.parent.config.output_dir
+            )
+            self.model.to(self.parent.config.device)
+
+            # Evaluate
+            predictions = await self._custom_accuracy(
+                eval_examples, features, dataset
+            )
+            results = squad_evaluate(eval_examples, predictions)
+
+        logger.info("Results: {}".format(results))
+
+        # return results
+        return Accuracy(results["f1"])
+
+    async def predict(
+        self, records: AsyncIterator[Record]
+    ) -> AsyncIterator[Tuple[Record, Any, float]]:
+        if not os.path.isfile(
+            os.path.join(self.parent.config.output_dir, "pytorch_model.bin")
+        ):
+            raise ModelNotTrained("Train model before prediction.")
+
+        self.model = AutoModelForQuestionAnswering.from_pretrained(
+            self.parent.config.output_dir
+        )  # , force_download=True)
+        self.model.to(self.parent.config.device)
+        self.tokenizer = AutoTokenizer.from_pretrained(
+            self.parent.config.output_dir,
+            do_lower_case=self.parent.config.do_lower_case,
+        )
+        async for record in records:
+
+            example = SquadExample(
+                qas_id=record.key,
+                question_text=record.feature("question"),
+                context_text=record.feature("context"),
+                answer_text=record.feature("answer_text"),
+                start_position_character=record.feature("start_pos_char"),
+                title=record.feature("title"),
+                is_impossible=record.feature("is_impossible"),
+                answers=record.feature("answers"),
+            )
+            features, dataset = squad_convert_examples_to_features(
+                examples=[example],
+                tokenizer=self.tokenizer,
+                max_seq_length=self.parent.config.max_seq_length,
+                doc_stride=self.parent.config.doc_stride,
+                max_query_length=self.parent.config.max_query_length,
+                is_training=False,
+                return_dataset="pt",
+            )
+            prediction = await self._custom_accuracy(
+                [example], features, dataset
+            )
+            record.predicted("Answer", prediction, "Nan")
+            yield record
+
+
+@entrypoint("qa_model")
+class QAModel(Model):
+    """
+    Implemented using `HuggingFace Transformers <https://huggingface.co/transformers/index.html>`_ Pytorch based Models.
+    Description about pretrianed models can be found `here <https://huggingface.co/transformers/pretrained_models.html>`_
```

Comment:
We should really `literalinclude` the `parser.py` code in here and talk about how we're using it to convert the format of the dataset. It might go best at the end.